### PR TITLE
gh-91602: Add iterdump() support for filtering database objects.

### PR DESCRIPTION
**Pull request by felixxm**
- [x] Docs changes.
- [x] More tests.
- [x] Release notes.

* Issue: gh-91602

### CODE REVIEW

File: Modules/_sqlite/connection.c [hunk 1]
```diff
@@ -1983,7 +1983,7 @@ Returns iterator to the dump of the database in an SQL text format.
 [clinic start generated code]*/
 
 static PyObject *
-pysqlite_connection_iterdump_impl(pysqlite_Connection *self)
+pysqlite_connection_iterdump_impl(pysqlite_Connection *self, const char *filter)
```
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: You'll want Argument Clinic to generate the `filter` parameter for you:

```diff
diff --git a/Modules/_sqlite/connection.c b/Modules/_sqlite/connection.c
index 2d5a7a10ee..9bb6a0a60e 100644
--- a/Modules/_sqlite/connection.c
+++ b/Modules/_sqlite/connection.c
@@ -1979,6 +1979,9 @@ pysqlite_connection_interrupt_impl(pysqlite_Connection *self)
 /*[clinic input]
 _sqlite3.Connection.iterdump as pysqlite_connection_iterdump
 
+    *
+    filter: object = None
+
 Returns iterator to the dump of the database in an SQL text format.
 [clinic start generated code]*/
 
```

After this change, you can do `make clinic`, which will regenerate `Modules/_sqlite/clinic/connection.c.h` correctly for you.

File: Modules/_sqlite/connection.c [hunk 2]
```diff
@@ -1998,8 +1998,9 @@ pysqlite_connection_iterdump_impl(pysqlite_Connection *self)
         }
         return NULL;
     }
-
-    PyObject *retval = PyObject_CallOneArg(iterdump, (PyObject *)self);
+    PyObject *py_filter = NULL;
+    py_filter = PyUnicode_FromString(filter);
+    PyObject *retval = PyObject_Call(iterdump, (PyObject *)self, py_filter);
```
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: You can use vector call for this (you'll find examples for how to do that a few other places in `connection.c`, IIRC).

File: Modules/_sqlite/connection.c (hunk 1 above)
Comment (felixxm) [reply Modules/_sqlite/connection.c]: Many thanks! Now I understand the workflow.

**Comment by erlend-aasland** [comment]
I see that the check-if-generated-files-are-up-to-date check still fails; you need to do a full `make regen-all`, in order to generate the needed static string object for the `filter` keyword parameter.

**Comment by erlend-aasland** [comment]
> Thanks for another tip 🎁 I'm not sure why but it also generated some unrelated lines 🤔.

Yeah, that's strange; I'd rebase onto `main` and `make regen-all` again 😄

**Comment by erlend-aasland** [comment]
BTW, can you please use `git pull --noff main` when aligning with `main`, instead of force pushing? :)

File: Lib/sqlite3/dump.py [hunk 3]
```diff
@@ -32,15 +32,23 @@ def _iterdump(connection):
         yield('PRAGMA foreign_keys=OFF;')
     yield('BEGIN TRANSACTION;')
 
+    if filter:
+        # Return database objects which match the filter pattern.
+        filter_name_clause = 'AND "name" LIKE %s'
```
Comment (erlend-aasland) [inline Lib/sqlite3/dump.py]: I guess you meant a placeholder, right?

```suggestion
        filter_name_clause = 'AND "name" LIKE ?'
```

File: Modules/_sqlite/connection.c (hunk 2 above)
Comment (felixxm) [reply Modules/_sqlite/connection.c]: Finally figured out how to use `PyObject_Vectorcall()` :tada:

File: Lib/sqlite3/dump.py (hunk 3 above)
Comment (felixxm) [reply Lib/sqlite3/dump.py]: Yes, I used Django's style placeholders :smile: , corrected.

**Comment by felixxm** [comment]
Thanks for all your help, much appreciated. I will work on tests and docs later today.

File: Modules/_sqlite/connection.c [hunk 4]
```diff
@@ -1998,8 +2003,10 @@ pysqlite_connection_iterdump_impl(pysqlite_Connection *self)
         }
         return NULL;
     }
-
-    PyObject *retval = PyObject_CallOneArg(iterdump, (PyObject *)self);
+    PyObject *args[2] = {(PyObject *)self, filter};
+    PyObject *kwnames = PyTuple_New(1);
+    PyTuple_SET_ITEM(kwnames, 0, PyUnicode_FromString("filter"));
+    PyObject *retval = PyObject_Vectorcall(iterdump, args, 1, kwnames);
```
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: You need to add error handling for the `PyTuple_New` and `PyUnicode_FromString` calls.

Comment (felixxm) [reply Modules/_sqlite/connection.c]: Added, thanks :+1:

File: Modules/_sqlite/connection.c [hunk 5]
```diff
@@ -2005,10 +2005,26 @@ pysqlite_connection_iterdump_impl(pysqlite_Connection *self,
...
+    }
+    PyTuple_SET_ITEM(kwnames, 0, py_filter);
+
     PyObject *retval = PyObject_Vectorcall(iterdump, args, 1, kwnames);
     Py_DECREF(iterdump);
     return retval;
+
+error:
+    Py_DECREF(iterdump);
+    Py_DECREF(args);
```
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: Args is an array of PyObject's, so we cannot directly decref it :) We don't need to worry about the refcount for the incoming PyObject's to `pysqlite_connection_iterdump_impl()` (that is: `self` and `filter`).

```suggestion
```

File: Modules/_sqlite/connection.c [hunk 6]
```diff
@@ -1998,10 +2003,28 @@ pysqlite_connection_iterdump_impl(pysqlite_Connection *self)
...
+        goto error;
+    }
+    py_filter = PyUnicode_FromString("filter");
+    if (!py_filter) {
+        goto error;
+    }
+    PyTuple_SET_ITEM(kwnames, 0, py_filter);
 
-    PyObject *retval = PyObject_CallOneArg(iterdump, (PyObject *)self);
+    PyObject *retval = PyObject_Vectorcall(iterdump, args, 1, kwnames);
```
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: We also need to decref `kwnames` and `py_filter` before returning on the "happy path".

Comment (felixxm) [reply Modules/_sqlite/connection.c]: `PyTuple_SET_ITEM` steals a reference to the `py_filter` so it seems there is no need to decref :thinking: 

https://docs.python.org/3/c-api/tuple.html#c.PyTuple_SET_ITEM

**Comment by felixxm** [comment]
Now it's ready for review :tada:

File: Modules/_sqlite/connection.c (hunk 6 above)
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Ah, yes, `PyTuple_SET_ITEM` is one of the weird ones; good call! :)

File: Modules/_sqlite/connection.c [hunk 7]
```diff
@@ -1998,10 +2003,28 @@ pysqlite_connection_iterdump_impl(pysqlite_Connection *self)
...
 
-    PyObject *retval = PyObject_CallOneArg(iterdump, (PyObject *)self);
+    if (!kwnames) {
+        goto error;
+    }
+    py_filter = PyUnicode_FromString("filter");
+    if (!py_filter) {
+        goto error;
+    }
+    PyTuple_SET_ITEM(kwnames, 0, py_filter);
```
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: Actually, let's use `Py_BuildValue` here; it will get the work done in very few lines :)

Comment (felixxm) [reply Modules/_sqlite/connection.c]: Great shortcut, thanks :confetti_ball:

**Comment by erlend-aasland** [comment]
Very nice; 321b3e0e03edfa6b5481e4856619b0ed41704679 made the job a lot easier. I still got some nitpicks, so bear with me :)

File: Modules/_sqlite/connection.c [hunk 8]
```diff
@@ -1998,10 +2003,21 @@ pysqlite_connection_iterdump_impl(pysqlite_Connection *self)
...
         return NULL;
     }
+    PyObject *args[2] = {(PyObject *)self, filter};
+    PyObject *kwnames = Py_BuildValue("(s)", "filter");
+    if (!kwnames) {
+        goto error;
+    }
 
-    PyObject *retval = PyObject_CallOneArg(iterdump, (PyObject *)self);
+    PyObject *retval = PyObject_Vectorcall(iterdump, args, 1, kwnames);
```
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: For the vectorcall, what you already wrote is perfectly fine, _but_ I'd prefer it if we followed the advice of PEP-590 and used `PyVectorcall_NARGS`; see other uses in connection.c for examples. This has implications for the `args` array and the values we pass to `PyObject_Vectorcall`.

**Review by erlend-aasland — COMMENTED** [review]
The refcout refactoring in 8cf51d42353fabf370aa8857b3ee8e9f6c14fbea is superb; I was just about to suggest it 🚀. I left a small comment regarding how we do the vector call.

When doing changes that affect reference counting, I normally do a refleak test locally:

```
$ ./python.exe -m test -R : test_sqlite3
```

File: Modules/_sqlite/connection.c (hunk 8 above)
Comment (felixxm) [reply Modules/_sqlite/connection.c]: `PyVectorcall_NARGS` is not widely used, but I tried to add it.

File: Modules/_sqlite/connection.c [hunk 9]
```diff
@@ -1998,9 +2003,16 @@ pysqlite_connection_iterdump_impl(pysqlite_Connection *self)
...
-
-    PyObject *retval = PyObject_CallOneArg(iterdump, (PyObject *)self);
+    PyObject *args[2] = {(PyObject *)self, filter};
+    PyObject *kwnames = Py_BuildValue("(s)", "filter");
+    if (!kwnames) {
+        Py_DECREF(iterdump);
+        return NULL;
+    }
+    Py_ssize_t nargs = PyVectorcall_NARGS(1);
+    PyObject *retval = PyObject_Vectorcall(iterdump, args, nargs, kwnames);
```
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: The following adjustments should do the trick (sorry if I pointed you in the wrong direction in the vector call PEP!):

```suggestion
    PyObject *args[3] = {NULL, (PyObject *)self, filter};
    PyObject *kwnames = Py_BuildValue("(s)", "filter");
    if (!kwnames) {
        Py_DECREF(iterdump);
        return NULL;
    }
    Py_ssize_t nargsf = 1 | PY_VECTORCALL_ARGUMENTS_OFFSET;
    PyObject *retval = PyObject_Vectorcall(iterdump, args+1, nargsf, kwnames);
```

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: See https://peps.python.org/pep-0590/#py-vectorcall-arguments-offset

**Review by erlend-aasland — COMMENTED** [review]
After this final change, we should be ready to land.

File: Modules/_sqlite/connection.c (hunk 9 above)
Comment (felixxm) [reply Modules/_sqlite/connection.c]: Changed.

**Comment by felixxm** [comment]
@erlend-aasland Thanks for all your help :gift:

File: Doc/library/sqlite3.rst [hunk 10]
```diff
@@ -1137,12 +1137,19 @@ Connection objects
...
       Useful when saving an in-memory database for later restoration.
       Similar to the ``.dump`` command in the :program:`sqlite3` shell.
 
+      :param entrypoint:
+
+        An optional ``LIKE`` pattern for database objects to dump, e.g. ``prefix_%``.
+        If ``None`` (the default), all database objects will be included.
+
+      :type entrypoint: str | None
+
```
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]: Can you fix the param name here?

File: Misc/NEWS.d/next/Library/2024-01-24-20-51-49.gh-issue-91602.8fOH8l.rst [hunk 11]
```diff
@@ -0,0 +1,3 @@
+Added ``filter`` keyword-only parameter to
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2024-01-24-20-51-49.gh-issue-91602.8fOH8l.rst]: Please use stars to mark up param names. Also, I think we should add an entry to What's New. Reusing the wording from the NEWS entry is normally ok.

**Review by erlend-aasland — COMMENTED** [review]
Now for the docs; I left some comments :)

File: Doc/library/sqlite3.rst (hunk 10 above)
Comment (felixxm) [reply Doc/library/sqlite3.rst]: Silly me :sweat_smile:

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
!buildbot .*refleak.*

**Comment by erlend-aasland** [comment]
The `buildbot/s390x Fedora Rawhide Refleaks PR` buildbot failure is unrelated to this PR.

**Comment by erlend-aasland** [comment]
Thanks a lot, @felixxm!

**Pull request merged**